#Capítulo 1 Fundamentos de redes neuronales usando Tensorflow 2.0

#Introducción.

## Tensorflow 
Es un software (framework) de código abierto de Deep Learning desarrollado por google,¡.

## Keras
Es una API de alto nivel, para entrenamiento y construcción de modelos de aprendizaje profundo.

##Redes Neuronales Artificiles ANNs

representan un clase de modelos de aprendizaje máquina, inspirada por estudios sobde el sistema nervioso central de los mamíferos. Cada ANN esta construida de muchas neuronas interconectadas, organizadas por capas. Neuronas en una capa pasan mensajes a neuronas en la siguiente capa y es asi como las neuronas calculan cosas.

Estudios iniciales que empezaron en los 50s con la introducción del perceptron, una red de dos capas usada para operaciones simples, y posteriormente expandida in los 60s con la introduccion de "back-propagation" algoritmo usado por si eficiente entrenamiento multi-capa.

#Perceptron.

El perceptron es un algoritmo simple, dada la entrada de un vector $x$ de $m$ valores $(x_1,x_2, ... , x_m)$, usualmente llamdas como características de entrada, o solamente caracteristicas, la salida es o un 1("Si") o un 0 ("No")


$f(x) = 1, wx + b > 0$

$f(x) = 0,$ de otra forma

donde:

$w$: es un vector de pesos, 
$wx$: es el producto punto de $\sum_{j=1}^{m}{w_jx_j}$
$b$: bias.

Notar que no puede expresar una respuesta de "tal vez".



## 1.   Red neuronal perceptrón



In [12]:
import tensorflow as tf
W = tf.Variable(tf.ones(shape=(2,2)), name="W")
b = tf.Variable(tf.zeros(shape=(2)), name="b")

@tf.function
def model(x):
	return (W*x+b)
out_a = model([1,0])

print(out_a)


tf.Tensor(
[[1. 0.]
 [1. 0.]], shape=(2, 2), dtype=float32)



## Redes neuronales multicapa. (MLP)

Modelo que cuenta con mutiples capas lineales, las capas de entrada y salida son visibles desde el exterior, mientras que el resto (capas intermedias) son ocultas.

A continuación se presentra un ejemplo de una MLP, para poder detectar numeros de imagenes con numeros escritos a mano.



importamos las bibliotecas

In [13]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

Descripción y parametros de entrenamiento.

In [14]:
EPOCHS = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # números de salidas = número de dígitos
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # Entrenamiento reservado para validación

Cargando base de datos de MNIST.

In [15]:
# Se puede verificar que la separación entre entrenamiento y evaluación es de 60,000, y
#10,000 respectivamente.
# Etiquetas
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()

Se modifican las dimensiones de los arreglos

In [16]:
# X_train son 60000 filas con valores de tamaño 28x28; reestructuramos a
# 60000 x 784.
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

Se normaliza las entradas

In [17]:
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples') #imprime la cantidad de muestras de entrenamiento
print(X_test.shape[0], 'test samples') #imprime las muestras que se evaluaran

60000 train samples
10000 test samples


Representación de una variable ficticia de etiquetado.

In [18]:
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)

Se construye el modelo

In [19]:
#softmax es la función sigmoidea, que active de manera suave la siguiente capa

model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(NB_CLASSES, input_shape=(RESHAPED,), name='dense_layer', activation='softmax')) 

Compilación de la red.

In [20]:
model.compile(optimizer='SGD', loss='categorical_crossentropy', metrics=['accuracy'])

Entrenamiento de la red.

In [21]:
model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=EPOCHS, verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

Epoch 1/200
375/375 [==============================] - 1s 2ms/step - loss: 1.7878 - accuracy: 0.4725 - val_loss: 0.9046 - val_accuracy: 0.8276
Epoch 2/200
375/375 [==============================] - 1s 2ms/step - loss: 0.8570 - accuracy: 0.8182 - val_loss: 0.6614 - val_accuracy: 0.8595
Epoch 3/200
375/375 [==============================] - 1s 2ms/step - loss: 0.6682 - accuracy: 0.8485 - val_loss: 0.5646 - val_accuracy: 0.8707
Epoch 4/200
375/375 [==============================] - 1s 2ms/step - loss: 0.5860 - accuracy: 0.8597 - val_loss: 0.5107 - val_accuracy: 0.8802
Epoch 5/200
375/375 [==============================] - 1s 2ms/step - loss: 0.5340 - accuracy: 0.8698 - val_loss: 0.4763 - val_accuracy: 0.8838
Epoch 6/200
375/375 [==============================] - 1s 2ms/step - loss: 0.5052 - accuracy: 0.8733 - val_loss: 0.4518 - val_accuracy: 0.8882
Epoch 7/200
375/375 [==============================] - 1s 2ms/step - loss: 0.4839 - accuracy: 0.8734 - val_loss: 0.4335 - val_accuracy: 0.8908

Evaluación de la red

In [22]:
test_loss, test_acc = model.evaluate(X_test, Y_test) 
print('Test accuracy:', test_acc)


313/313 [==============================] - 0s 919us/step - loss: 0.2771 - accuracy: 0.9225
Test accuracy: 0.9225000143051147


#Conclusión.
El modelo tiene buen desempeño al darnos un 92% de precisión.Sin embargo esto puede mejorar o emperoar dependiendo de los parametros de configuración con los que se forma el modelo, por lo tanto el resultado depende bastante de los parametros y para poder obtener los mejores es necesario jugar con diferentes configuraciones para obtener una precisión aceptable.